consensus/src/experimental/buffer_manager.rs

```diff
@@ -204,33 +204,45 @@ impl StateManager {
-    fn try_persisting_blocks(&mut self) {
+    /// Pop the prefix of buffer items until (including) aggregated_item_cursor
+    /// Prepare an aggregated item and send it to the persisting phase
+    fn try_persisting_blocks(&mut self, aggregated_item_cursor: BufferItemRootType) {
         let mut cursor = self.buffer.head.as_ref().cloned();
+        let mut blocks_to_persist: Vec<Arc<ExecutedBlock>> = vec![];
+
         while cursor.is_some() {
             let buffer_item = take_elem(&cursor);
-            if let BufferItem::Aggregated(aggregated) = buffer_item {
-                let blocks: Vec<Arc<ExecutedBlock>> = aggregated
-                    .executed_blocks
-                    .into_iter()
-                    .map(Arc::new)
-                    .collect();
-                // send to persisting phase
-                block_on(self.persisting_phase_tx.send(PersistingRequest {
-                    blocks,
-                    commit_ledger_info: aggregated.aggregated_proof,
-                    callback: aggregated.callback,
-                }))
-                .ok();
-                cursor = get_next(&cursor);
-                self.buffer.pop_front();
-            } else {
-                // we put the item back
-                set_elem(&cursor, buffer_item);
+            blocks_to_persist.extend(
+                buffer_item
+                    .get_blocks()
+                    .iter()
+                    .map(|eb| Arc::new(eb.clone()))
+                    .collect::<Vec<Arc<ExecutedBlock>>>(),
+            );
+            if link_eq(&cursor, &aggregated_item_cursor) {
+                if let BufferItem::Aggregated(aggregated) = buffer_item {
+                    // send to persisting phase
+                    block_on(self.persisting_phase_tx.send(PersistingRequest {
+                        blocks: blocks_to_persist,
+                        commit_ledger_info: aggregated.aggregated_proof,
+                        // we use the last callback
+                        // this is okay because the callback function (from BlockStore::commit)
+                        // takes in the actual blocks and ledger info from the state computer
+                        // the encoded values are references to the block_tree, storage, and a commit root
+                        // the block_tree and storage are the same for all the callbacks in the current epoch
+                        // the commit root is used in logging only.
+                        callback: aggregated.callback,
+                    }))
+                    .ok();
+                } else {
+                    panic!("Bad Aggregated Item Cursor: Item is not aggregated.");
+                }
                 break;
             }
+            cursor = get_next(&cursor);
+            self.buffer.pop_front();
         }
+        unreachable!("Bad Aggregated Item Cursor: Item not found in the buffer.");
```

Comment:
why this is unreachable?

Reply:
This function assumes the aggregated item exists in the buffer.

Reply:
it only `break` after sending the request instead of `return`

Reply:
oops sorry I meant to return after sending the request